### ISSUE TITLE
Font Size and Font Name drop-downs do not always reflect font styling

### ISSUE THREAD

**Issue by jswiderski** [ckeditor/ckeditor-dev#1973]
## Type of report

Feature request

## Provide description of the new feature

Please see: https://dev.ckeditor.com/ticket/9964
It would be nice if dropdowns would show any font name or size used in the editor.

At the moment dropdowns only reflect font size and font name that exactly matches the font styles from dropdown. If there is a different tag used (e.g. font style is assigned directly `div` or `font` instead of `span` is used ), or even `font-family` content is slight different to what is defined in dropdown, there will be no match. It would be nice if dropdown on the toolbar could be some sort of preview for any type of font used.

**Comment by jswiderski** [comment]
One problem which I personally have with this feature is that I'm not sure how it really should work.

Dropdown Button on Toolbar can be treated as a sort of information label which displays font name and size whenever selection changes (could be obtained from `style` or `getComputedStyle`) and this is all fine but: 

1. If there is a custom font used which is not defined for dropdown, user will see it on the toolbar but will not see it selected (or rather not see it at all) after opening the dropdown. Is this ok or should perhaps such style be added to dropdown dynamically?  While this sounds nice it may lead to bloating the dropdown in case of large documents with many different fonts used (this is rather rare but can happen).
2. If there is a font defined in the dropdown e.g. `Comic Sans` but it is assigned directly to e.g. `div` or `font` tag, there will be no match as well because font style definition requires `span`. This will be definitely considered by end users as a bug because they will see `Comic Sans` on Toolbar but they will not see it selected in dropdown. In case of `font` you might say this is not a big issue because in worst case scenario you will need to apply `Comic Sans`  twice in order to disable it. First to change `font` tag to `span` (dropdown can do that already) and second to remove the `span` once definition is correct. The bigger problem is when we are dealing with font style assigned to `div` directly (we are talking about case pasted through source mode and ACF disabled what is not that uncommon unfortunately). Not only we won't be able to change that font on `div` but by applying a style we will end up with `div` holding inner `span` and both having same font assigned (or different is user decides to change the font). This may lead to weird styling issues reported by users and it may happen that information label will bring more confusion than it brings information because it is natural that once you have information in dropdown, you expect dropdown to follow and work as you would expect it meaning removing font from `div` and applying inner `span` but this would be a major makeover.

Having control over what user enters or pastes would be a huge help but if user has source mode, ACF disabled and pastes external HTML, that control is gone. There are also browser weird bugs like e.g. 
1. Use this HTML and selection: `<p><span style="font-family:Lucida Sans Unicode,Lucida Grande,sans-serif"><span style="font-size:72px">test^</span></span></p>`
2. Start deleting characters one by one. Once you delete everything, start typing. 
3. Result `font` tag will be applied out of nowhere. In such a case even if we have full control over what user enters, there is nothing we can do as this particular case is like pasting external HTML with ACF disabled beyond any editor control.

**Comment by marco-assentis** [comment]
I agree with all your doubts if you take as possible input to the editor the existing and future universe of possible html/css. You will not find a satisfying solution. Also I disagree that this should be the responsibility of the drop down to handle all cases. In my opinion **it should be the duty of the editor as a whole to define what content it can _reliable_ handle and what not**. In some cases plugins may workaround issues where the editor cannot do so due technical limitations.

Try to see it from the users perspective
There are multiple types of users (probably there are more, but this are the 2 we are confronted with). 
1) User with understanding of html/css. 
2) and users without any understanding relying on the editor.

To cover all the cases for type 1) is not feasible. But also these users are not so concerned about such functionality as they hack in the source anyways if the editor does not do or show on what they want. So In my opinion not so important.

User 2) is much more concerned about such functionality. For this user, there is no way other than the editor toolbar to find out, what font is used. If the user sees formatted text, but the dropdown shows "default", this is simply confusing. So try to focus on this user. What does this user do? He creates content with the editor. So you control on the markup which is created! That is already the current state of this drop down. If the font is applied using <span style="font-family: MyFont"/> it will show it as expected. Now there are some enemies to that approach. And why it is not finished and from my point of view rather a BUG than a feature request.

a) The Browser. As you describe, If the user enters text, marks it, applies font family and or size, then deletes the text again with backspace, the font settings are reapplied from the browser when he starts typing again. BUT the browser does it using the <font> tag and the size unit is always px. For the user, this looks initially like his style have been applied properly and he is happy typing. Next day he wants to check his font settings he did yesterday and selects the text. Hmmmm no settings applied, but it looks right... wtf. So this case can be handled in multiple ways. Either the editor is smart enough to convert such cases when they appear (I heard that is not possible for some reason) OR, the drop down supports other kinds of font style. Now in addition it just needs to support <font> and the ability to concert px to the configured font-size unit in case of font-size. User happy. 

b) Copy & Paste: This is a nasty one as any content might be pasted. But hey, you can filter that. Simply remove the unsupported tags. Maybe the user will complain that some formatting may not be taken over, but this is still better than uncontrolled content in the editor, as this will confuse ALL plugins. Nice to have would than be the ability to convert SOME input to the expected input html... But that is already next step.

c) setData and getData APIs. With setData an application might set any content. But you can also filter that like copy & paste and document what is supported. If you really want you can still have a "freestyle mode" where you allow everything. But than don't be surprised it the plugins cannot deal with it.
On getData() it would be nice if browser added tags would already be converted, such that you have a clean interface. 

d) Source View Plugin: Well, needless to say that this is the domain of user 1) so no need to handle.

With such an approach you could satisfy at least the users which actually require a free text editor to create content. You simply document the limitations for the other type of user and everybody would be fine. At the moment, Type 2 User is confused and gets frustrated. Type 1 user just hacks the source view. Also in our case there are 1000s of Type 2 users vs maybe 100 Type 1 user.

Maybe that helps in seeing the issue from a different perspective. Looking forward to hear your thoughts on this.

**Comment by jswiderski** [comment]
I could not disagree more whit tha you have written.

> If  the user sees formatted text, but the dropdown shows "default", this is simply confusing.

CKEditor dropdowns have never worked that way so user has no reason to be confused in the first place. How can a user expect functionality that was never there?

> BUT  the browser does it using the tag and the size unit is always px.

That is incorrect. Edge for example will use (I'm talking about borwser inserting `font` tag by itself) something like `<font face="Comic Sans MS" size="7">ddddd</font>` where size 7 will be used for `72`, `48` and `36px`. How will you convert that to a proper unit?

> Next day he wants to check his font settings he did yesterday and selects the text. Hmmmm no settings applied, but it looks right... wtf. 

 If this was just about saving and checking font the next day there would be no issue here as ACF will convert `font` to `span` tag once you switch to source mode or get data from editor. Yes ACF can do that.
The problem here is checking the font in real time, right after the browser has done its weird magic trick. In Browser `font` tag quirk, the browser changes HTML and CKEditor does not know anything about it. This is the same like you opening the development tools and changing HTML in the editor from there. CKEditor will not react in such case as well. HTML is being changed through backdoor and CKEditor can't do anything about it. 

> With setData an application might set any content. But you can also filter that like copy & paste and document what is supported. 

All the things you have mentioned at the end of your email are already there. We have ACF, paste filters and `afterPasteFromWord` event to extra modify content pasted from Word. There is also possibility to remove Source Mode plugin and this all makes CKEditor a controlled environment. With such a setup the feature you are asking about is already there because dropdown will react correctly to styles created only through that dropdown. Additionally you can use the editor API to control fonts to a small subset having guarantee that correct font will always be shown in the dropdown - https://codepen.io/j_swiderski/pen/mqYbrr.  The browser quirk with font tag changes all that to user 1 scenario (in this particular case) because as I have already explained – this is like inserting HTML through dev-tools and there are no fancy mechanisms to defend against that.

 This feature is no about showing font styles in controlled environment because we have that already. This is about showing information for any font used and font used in any type of element. Both Browser Quirk as well as having ACF and paste filters disabled plus pasting invalid HTML into source mode (and believe me this is not that uncommon) fall into this category. In such cases it is not easy to find a good solution and we are not interested in introducing any semi solution that will work only in small subset of cases.

**Comment by marco-assentis** [comment]
> CKEditor dropdowns have never worked that way so user has no reason to be confused in the first place. How can a user expect functionality that was never there?

2 screenshots from https://ckeditor.com/ckeditor-4/#document

1. Marke Text, choose font family "Comic Sans MS", mark a word, check what the dropdown shows => "Comic Sans MS"
<img width="878" alt="step_1" src="https://user-images.githubusercontent.com/26218819/40143642-4963ec1c-595c-11e8-8317-c7a788610514.png">

2. Delete all the text of that sentence with backspace until the cursor is just before there where the first character was. Retype "Please find below...", mark a word, check what the drop down says => "Default"
<img width="871" alt="step_2" src="https://user-images.githubusercontent.com/26218819/40143660-545fdee6-595c-11e8-8321-94c03e283365.png">

> CKEditor dropdowns have never worked that way

For the user, CKEditor just worked seconds ago that way. Not even for a developer this is obvious in the first place until he traced down what the browser does.

With the ACF etc. I agree and that is also what we are using. 

> his feature is not about showing font styles in controlled environment because we have that already. 

No, you have it just in 95% stage. The tricky parts dealing with browser magic you miss. And this makes the editor finally unpredictable for end users.
